### ISSUE TITLE
v6 Does Not Allow Journaling of RunAlways Scripts

### ISSUE THREAD

**Issue by champin4** [DbUp/DbUp#814]
I understand this was a decision and labeled as a breaking change, but our company really needs to log every script that is run - no matter how many times it is run.

Proposed solutions are:

1. Make the ScriptExecutor.ShouldJournalScript() virtual so it can be overridden 
2. Add a public bool that determines if only RunOnce scripts will be logged and use it in the private ShouldJournalScript() logic. Then it could be overridden during configuration setup

Open to other solutions as well

**Comment by droyad** [comment]
As per #789, another solution is 

The upgrader returns a DatabaseUpgradeResult which has a collection of SqlScript (.Scripts) that were run. I suggest logging or recording these. For auditability this may be a better solution anyway as you can include other data as well (e.g. deployment it was part of, who ran it, and other log output).

**Comment by droyad** [comment]
I'll leave this open for comment though. I can see the value in recording it. Personally I'm in favour of making the behaviour of the journaling extensible instead of adding a knob.

**Comment by drewburlingame** [comment]
Hi, I've taken over maintenance of the [dbup-cli](https://github.com/drewburlingame/dbup-cli) project.  An [issue we have](https://github.com/drewburlingame/dbup-cli/issues/8) is that `IsUpgradeRequired` always returns true if there is a `RunAlways` script, even if that script is has already been applied.  If the most common use case for `RunAlways` is to create functions, stored procedures and the like, then it would be ideal to only apply the scripts if there are changes and have `IsUpgradeRequired` return false when not.  This would require journaling and also storing a hash to identify changes.

**Comment by champin4** [comment]
Yes, we have a real world use case where we want to know what has been run - but we also don't feel the need to run a script every time if it has not changed (e.g. a stored procedure).

We called them RunAnytime scripts and leverage hashing like you mentioned. We also use hashing for the RunOnce scripts to guard against any errant changes to those scripts.

So our RunAnytime method that we developed was what drove this ticket creation, since RunAlways scripts are no longer journaled we can't turn them into our "RunAnytime" because we can't compare the hashes stored in the journal (without even more extending, which we could do except the solution to making journaling RunAlways scripts optionally journal is simple)

**Comment by champin4** [comment]
Agree on generally making extensible vs. a knob. But the code itself to remove the RunAlways journaling simply introduced a knob forced to off (false).

**Comment by drewburlingame** [comment]
I too agree generally about making extensible vs a knob.  I would add that if there is a common way in which a set of users extend the library, then the effort is duplicated until someone creates an extension package, at which point it becomes a knob plus a nuget dependency.  And then that knob can fall behind increasing the risk for users to adopt it.

**Comment by drewburlingame** [comment]
@fr4gles I'm not a maintainer of this library.  I'm just proposing they include the knob in this case since I think both options are common use cases.
